### PR TITLE
Feat: add support for Provider-defined functions

### LINKED ISSUE #199: Add support for Provider-defined functions
1. Add test to cover the new use-case.
2. Upgrade terratag dependencies.

### PR DESCRIPTION
**Pull request by TomerHeber**
1. Updated to go 1.23.
2. Upgraded dependencies.
3. Added a test for provider function.
4. Fixed google tests for older versions (removed tests).
5. Removed current linter, replacing with golang lint ci.

fixes #199

### CODE REVIEW

File: .golangci.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+linters:
```
Comment (TomerHeber) [inline .golangci.yaml:1]: new linter... the older one is longer active...

File: go.mod [hunk 2]
```diff
@@ -1,31 +1,30 @@
 module github.com/env0/terratag
 
-go 1.21
+go 1.23
 
 require (
 	github.com/bmatcuk/doublestar v1.3.4
 	github.com/hashicorp/go-hclog v1.5.0
-	github.com/hashicorp/hcl/v2 v2.16.2
+	github.com/hashicorp/hcl/v2 v2.22.0
```
Comment (TomerHeber) [inline go.mod:8]: upgraded the versions to support provider functions.

File: test/fixture/terraform_12/config.yaml [hunk 3]
```diff
@@ -10,5 +10,4 @@ suites:
   - git_issue_56__line_ends_with_closing_paren
   - git_issue_66_aws_aliased_provider
   - git_issue_88__corrupt_hcl_file
-  - google_beta_resource_skip
-  - google_vs_google_beta
\ No newline at end of file
+  - google_beta_resource_skip
```
Comment (TomerHeber) [inline test/fixture/terraform_12/config.yaml:13]: removed this test from older providers where setting a specific provider version doesn't work...

File: test/tests/google_vs_google_beta/expected/main.tf [hunk 4]
```diff
@@ -3,7 +3,8 @@
 terraform {
   required_providers {
     google = {
-      source = "hashicorp/google"
+      source  = "hashicorp/google"
+      version = "4.84.0"
```
Comment (TomerHeber) [inline test/tests/google_vs_google_beta/expected/main.tf:7]: need to lock the version... this is because the 'labels' was removed from beta state (now in main version).
therefore, need to lock to a version where this was beta.

File: internal/convert/terratag_locals.go [hunk 5]
```diff
@@ -34,11 +34,13 @@ func decodeTerratagLocals(locals Locals, s string) error {
 func encodeTerratagLocals(locals Locals) string {
 	ret := "{"
 
-	// Return it in ordered manner for order consistency (primarly when running tests).
-	var keys []string
+	// Return it in ordered manner for order consistency (primarily when running tests).
+	keys := []string{}
```
Comment (sagydr) [inline internal/convert/terratag_locals.go:38]: Is this related to the versions upgrade?

File: internal/tag_keys/tag_keys.go [hunk 6]
```diff
@@ -1,8 +1,9 @@
 package tag_keys
 
 import (
-	"github.com/hashicorp/hcl/v2/hclwrite"
 	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclwrite"
```
Comment (sagydr) [inline internal/tag_keys/tag_keys.go:6]: what's the change here?

File: terratag.go [hunk 7]
```diff
@@ -211,34 +223,40 @@ func tagFileResources(path string, args *common.TaggingArgs) (*counters, error)
...
 			}
+
 			tagsMap[match[1]] = match[2]
 		}
 	}
 
 	keys := utils.SortObjectKeys(tagsMap)
 
-	var mapContent []string
+	mapContent := []string{}
```
Comment (sagydr) [inline terratag.go:255]: How come you don't need the `var` here?

File: test/tests/provider_function/expected/main.terratag.tf [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "5.68.0"
+    }
+  }
+}
+
+resource "aws_ecr_repository" "hashicups" {
+  name = "hashicups"
```
Comment (sagydr) [inline test/tests/provider_function/expected/main.terratag.tf:11]: Why is it called `hashicups` ? related to the [hashicups provider?](https://registry.terraform.io/providers/hashicorp/hashicups/latest/docs)

File: terratag.go (hunk 7 above)
Comment (TomerHeber) [reply terratag.go:255]: ahh good question!
`:=` is shortcut that does not need var.

File: test/tests/provider_function/expected/main.terratag.tf (hunk 8 above)
Comment (TomerHeber) [reply test/tests/provider_function/expected/main.terratag.tf:11]: I just took the example from the offical website...

File: internal/convert/terratag_locals.go (hunk 5 above)
Comment (TomerHeber) [reply internal/convert/terratag_locals.go:38]: As part of the upgrade I had to replace the linter. So there are linter fixes as well.

File: internal/tag_keys/tag_keys.go (hunk 6 above)
Comment (TomerHeber) [reply internal/tag_keys/tag_keys.go:6]: automattic golang formatting... nothing changed... I guess the formatter fixed it this way...

**Review by sagydr — APPROVED** [review]
